### ISSUE TITLE
Unpool with indices

### ISSUE THREAD

**Issue by xianlopez** [dvazquezcvc/mcv-m5#32]
In class we have been told that in layers/ourlayers.py there is the layer to perform max unpool with indices, as required by SegNet. However, I can't find it. In layer DePool2D, I don't see how to pass the indices...

**Comment by dvazquezcvc** [comment]
Hi,

The Depool layer expects the maxpooling layer that you want to unpool.

# Downsample blocks of the basic-segnet
def downsampling_block_basic(inputs, n_filters, filter_size,
                             W_regularizer=None):
    # This extra padding is used to prevent problems with different input
    # sizes. At the end the crop layer remove extra paddings
    pad = ZeroPadding2D(padding=(1, 1))(inputs)
    conv = Convolution2D(n_filters, filter_size, filter_size,
                         border_mode='same', W_regularizer=W_regularizer)(pad)
    bn = BatchNormalization(mode=0, axis=channel_idx())(conv)
    act = Activation('relu')(bn)
    maxp = MaxPooling2D(pool_size=(2, 2), strides=(2, 2))(act)
    return maxp


# Upsample blocks of the basic-segnet
def upsampling_block_basic(inputs, n_filters, filter_size, unpool_layer=None,
                           W_regularizer=None, use_unpool=True):
    if use_unpool:
        up = DePool2D(unpool_layer)(inputs)
    else:
        up = UpSampling2D()(inputs)
    conv = Convolution2D(n_filters, filter_size, filter_size,
                         border_mode='same', W_regularizer=W_regularizer)(up)
    bn = BatchNormalization(mode=0, axis=channel_idx())(conv)
    return bn


# Create model of basic segnet
def build_segnet_basic(inputs, n_classes, depths=[64, 64, 64, 64],
                       filter_size=7, l2_reg=0.):
    """ encoding layers """
    enc1 = downsampling_block_basic(inputs, depths[0], filter_size, l2(l2_reg))
    enc2 = downsampling_block_basic(enc1, depths[1], filter_size, l2(l2_reg))
    enc3 = downsampling_block_basic(enc2, depths[2], filter_size, l2(l2_reg))
    enc4 = downsampling_block_basic(enc3, depths[3], filter_size, l2(l2_reg))

    """ decoding layers """
    dec1 = upsampling_block_basic(enc4, depths[3], filter_size, enc4,
                                  l2(l2_reg))
    dec2 = upsampling_block_basic(dec1, depths[2], filter_size, enc3,
                                  l2(l2_reg))
    dec3 = upsampling_block_basic(dec2, depths[1], filter_size, enc2,
                                  l2(l2_reg))
    dec4 = upsampling_block_basic(dec3, depths[0], filter_size, enc1,
                                  l2(l2_reg))

    """ logits """
    l1 = Convolution2D(n_classes, 1, 1, border_mode='valid')(dec4)
    score = CropLayer2D(inputs, name='score')(l1)
    softmax_segnet = NdSoftmax()(score)

    # Complete model
    model = Model(input=inputs, output=softmax_segnet)

    return model
